Improve LSP1-UniversalReceiver + Delegate standard explanation (#404)
lukso-network/docs: docs/standards/generic-standards/lsp1-universal-receiver.md

:::

## Introduction
There is often the need for smart contracts to **be aware of incoming transactions**, especially when it comes to token transfers.

> A good example is **[ERC20](https://eips.ethereum.org/EIPS/eip-20)** token transfers. When a smart contract receives a token, it has no generic way **to be notified** about it. As the ERC20 token contract acts like a registry and during the transfer, the sender's balance is decreased, and the recipient's balance is increased **without further interaction**.

Reviewer: _Suggesting splitting in multiple sentences_

```suggestion
A good example is **[ERC20](https://eips.ethereum.org/EIPS/eip-20)** token transfers. When a smart contract receives a token, it has no generic way **to be notified** about it. During the token transfer, the sender's balance decreases, and the recipient's balance increases. There are **no further interactions**, and the ERC20 token contract just acts as a registry.
```

- bytes32 `typeId`: Hash or Hook of a specific standard.
- bytes `data`: Any arbitrary data.

> Receiving contracts should consider the `typeId` parameter to **decode the data correctly**.

The `universalReceiver(...)` function **emits an event with the data passed to it and some additional data**. 

![universalReceiver function emits an event](/img/standards/lsp1/universal-receiver-event.jpeg)

> For example, **to mark the receiving of a specific token**, during a token transfer, the token contract can call the `universalReceiver(..)` function of the token recipient with:

Reviewer: _Suggested changes_

```suggestion
For example, **to notify the recipient that he is about to receive tokens**, during a token transfer, the token contract can call the `universalReceiver(..)` function of the recipient with:
```

- bytes32 `typeId`: **Hash**('ERCXXXXTokenReceived')
- bytes `data`: **packedData**(amount of token sent, the sender address, the block timestamp)

In this way, instead of **listening to all the events of the token contrats on the network**, and checking which one of these transfers is relative to the recipient, users can listen to the **[UniversalReceiver](../smart-contracts/lsp0-erc725-account.md#universalreceiver-1)** event on the contract implementing the `universalReceiver(..)` and know the token transfer details.  

> Also, the function can then implement **custom logic** to make the contract behave differently based on the data received. For instance, the `universalReceiver(...)` function, if customized, can do the following:

Reviewer: _Suggesting shortening sentence

```suggestion
The `universalReceiver(...)` function implement **custom logic** to make the contract behave differently based on the data received. Some ideas include:
```

- Reverting on calls to completely disallow the smart contract from receiving assets, information, etc. :x:

> - Registering the received assets inside the contract storage (see [LSP5 - Received Assets](../universal-profile/lsp5-received-assets.md)). :heavy_plus_sign:

Reviewer: I suggest being as creative as possible here 🎨 🙂 , with multiple ideas to really explain **how extensible LSP1 is:**

- Disallow receiving specific tokens from specific token contract addresses, for instance (_e.g: spam tokens)._
- Forward and store specific tokens in a [LSP9 Vault]("link to LSP9 Standard").
- Stake-specific tokens in a staking contract (e.g: governance tokens).
- Forward specific tokens in a contract behind a protocol or dApp (e.g: liquidity or lending pool to earn interest).
- Depending on the `typeId`, save a percentage % of tokens received (native tokens or not), by placing them in an LSP9 Vault for instance.

Author: Good ideas, will not specify LSP9Vault as it could be overcomplicated for new users reading the standard, will keep it generic and write vaults

- Forwarding all the received assets to an external vault or a staking contract.

![universalReceiver function execute custom logic](/img/standards/lsp1/universal-receiver-logic.jpeg)

## Extension

:::info

See the **[LSP1-UniversalReceiverDelegate](../generic-standards/lsp1-universal-receiver-delegate.md)** standard for more details.

:::

> Customizing the `universalReceiver(..)` function is an option that users have to allow the function to **behave differently on different actions**. However, it's not advised to hardcode the logic of reacting on specific actions inside the function, as **this logic may need to change in the future** depending on several factors (eg. the vault where the tokens are forwarded get hacked, new staking contract is deployed, decided to revert on specific tokens later). 

Reviewer: _Suggested changes_

```suggestion
Overriding and customizing the `universalReceiver(..)` function is an option for users to allow **different behaviours depending on the data received**. However, it's not advised to hardcode the logic of reacting to specific actions inside the function because **this logic may need to change in the future** depending on several factors (eg. the vault where the tokens are forwarded gets compromised, a new staking contract is deployed, decided to revert on specific tokens later). 
```

**[LSP1-UniversalReceiverDelegate](../generic-standards/lsp1-universal-receiver-delegate.md)** is an **optional extension** to the **[LSP1-UniversalReceiver](#)** standard. As well as notifying a contract about the incoming and outgoing transactions by emitting an event, it can delegate the call to an external contract that can **handle and react to specific calls** with its custom logic.

![Universal Receiver Delegate contract](/img/standards/lsp1/universal-receiver-delegate.jpeg)

> The address of the **external contract** can be a variable that changes in the contract storage by a setter function. In this way, users can customize such contracts to implement a specific logic that is changeable anytime in the future.

Reviewer: _Suggested changes_

```suggestion
The address of the **external contract** can be stored and changed inside the contract storage. This way, users can customize such contracts to implement a specific logic that is changeable at any time.
```

![Multiple Universal Receiver Delegate](/img/standards/lsp1/multiple-urd.jpeg)

> If the contract implementing the `universalReceiver(..)` supports **[ERC725Y Data key-value store](https://github.com/ERC725Alliance/erc725/blob/main/docs/ERC-725.md#erc725y)**, the address of the **external contract** MUST be set as a value for the **LSP1UniversalReceiverDelegate data key** shown below to enable the optional extension. This key-value pair will act as a reference, making this external contract upgradeable if required.

Reviewer: Same changes suggested as above.

BUt if its's in two places and duplicated, maybe we can remove it in one of them?

Author: I would keep it, as people could read the 2 standards separately, it's not that big deal imo
